#Medical Insurance Cost Prediction

Health Insurance plays cruical part when ones health condition is imbalanced    this dataset helps to predict how much of insurance money will be required for a individual.

In [1]:
import pandas as pd
import numpy as np

In [2]:
mi=pd.read_csv('insurance.csv')
mi

age     sex   bmi  children smoker     region  expenses
0      19  female  27.9         0    yes  southwest  16884.92
1      18    male  33.8         1     no  southeast   1725.55
2      28    male  33.0         3     no  southeast   4449.46
3      33    male  22.7         0     no  northwest  21984.47
4      32    male  28.9         0     no  northwest   3866.86
...   ...     ...   ...       ...    ...        ...       ...
1333   50    male  31.0         3     no  northwest  10600.55
1334   18  female  31.9         0     no  northeast   2205.98
1335   18  female  36.9         0     no  southeast   1629.83
1336   21  female  25.8         0     no  southwest   2007.95
1337   61  female  29.1         0    yes  northwest  29141.36

[1338 rows x 7 columns]

Datsetset consist of 1338 rows and 7 variables amoung them expenses(medical expenses) is target variable and also dataset is nonnull

In [3]:
mi.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
expenses    0
dtype: int64

In [4]:
mi.drop_duplicates()

age     sex   bmi  children smoker     region  expenses
0      19  female  27.9         0    yes  southwest  16884.92
1      18    male  33.8         1     no  southeast   1725.55
2      28    male  33.0         3     no  southeast   4449.46
3      33    male  22.7         0     no  northwest  21984.47
4      32    male  28.9         0     no  northwest   3866.86
...   ...     ...   ...       ...    ...        ...       ...
1333   50    male  31.0         3     no  northwest  10600.55
1334   18  female  31.9         0     no  northeast   2205.98
1335   18  female  36.9         0     no  southeast   1629.83
1336   21  female  25.8         0     no  southwest   2007.95
1337   61  female  29.1         0    yes  northwest  29141.36

[1337 rows x 7 columns]

In [49]:
mi_copy=pd.read_csv('insurance.csv')
mi_copy.drop_duplicates()

age     sex   bmi  children smoker     region  expenses
0      19  female  27.9         0    yes  southwest  16884.92
1      18    male  33.8         1     no  southeast   1725.55
2      28    male  33.0         3     no  southeast   4449.46
3      33    male  22.7         0     no  northwest  21984.47
4      32    male  28.9         0     no  northwest   3866.86
...   ...     ...   ...       ...    ...        ...       ...
1333   50    male  31.0         3     no  northwest  10600.55
1334   18  female  31.9         0     no  northeast   2205.98
1335   18  female  36.9         0     no  southeast   1629.83
1336   21  female  25.8         0     no  southwest   2007.95
1337   61  female  29.1         0    yes  northwest  29141.36

[1337 rows x 7 columns]

In [5]:
mi[['sex','smoker','region']] = mi[['sex','smoker','region']].astype('category')
mi.dtypes

age            int64
sex         category
bmi          float64
children       int64
smoker      category
region      category
expenses     float64
dtype: object

Converted data from object type to categorical and converting categorical into numerical using Labelencoder

In [6]:
from sklearn.preprocessing import LabelEncoder
label= LabelEncoder()

In [7]:
label.fit(mi.sex)
mi.sex=label.transform(mi.sex)
label.fit(mi.smoker)
mi.smoker=label.transform(mi.smoker)
label.fit(mi.region)
mi.region=label.transform(mi.region)
mi.dtypes

age           int64
sex           int64
bmi         float64
children      int64
smoker        int64
region        int64
expenses    float64
dtype: object

In correlation(seaborn heatmap) target variable correlation is high only with 'smoker' and considerable with 'age' and 'bmi' others have significantly low correlation with target variable also ommiting 'region' variable since it has negative correlation with target variable(expenses)

In [53]:
 from sklearn.model_selection import train_test_split as tts
 from sklearn import metrics

 X=mi.drop(['expenses','region'],axis=1)
 y=mi['expenses']

 Xtrain, Xtest, ytrain, ytest = tts(X,y,test_size=0.3, random_state=0)

Random forest regressor used for predicting medical insurance cost of an individual, with mean_square_error as criterion

In [54]:
from sklearn.ensemble import RandomForestRegressor 
RandF_reg=RandomForestRegressor(n_estimators=100,criterion='mae',random_state=2,n_jobs=-1)
RandF_reg.fit(Xtrain,ytrain)

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mae',
                      max_depth=None, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=100, n_jobs=-1, oob_score=False,
                      random_state=2, verbose=0, warm_start=False)

In [55]:
X_train_pred = RandF_reg.predict(Xtrain)
X_test_pred = RandF_reg.predict(Xtest)

In [56]:
print('R2 train data: %.3f, R2 test data: %.3f' % (metrics.r2_score(ytrain,X_train_pred, ytrain),metrics.r2_score(ytest,X_test_pred, ytest)))

R2 train data: 0.967, R2 test data: 0.855


In [57]:
mse=metrics.mean_absolute_error(ytest,X_test_pred)
print('Mean absolute error:',mse)

Mean absolute error: 2712.629334577115


Above cells display metrics and r2_score yeilded by Random Forest Regresor

In [58]:
mi_comp = pd.DataFrame({'Actual':ytest,'predicted':X_test_pred})
mi_comp.head(8)

Actual    predicted
578    9724.53   9993.86580
610    8547.69   9144.37410
569   45702.02  44599.98460
1034  12950.07  13426.69060
198    9644.25  10338.13170
981    4500.34  13061.75690
31     2198.19   2192.77405
1256  11436.74  11337.85470

Comparision of actual result and predicted results as we can observer the difference is less

In [59]:
Xtrain.head(4)

age  sex   bmi  children  smoker
1163   18    0  28.2         0       0
196    39    0  32.8         0       0
438    52    0  46.8         5       0
183    44    0  26.4         0       0

Inputs
1. age(in years)int
2. sex(0 female, 1 male)
3. bmi(body mass index)float
4. children(number)int
5. smoker(0 if no, 1 if yes)


In [44]:
MI=RandF_reg.predict([[25,33,2,1,2]])
print(MI)

[35973.9365]
